### PR TITLE
client: return write success on <= K failures

### PR DESCRIPTION
**Pull request by andriytk**
The redundancy of the user data will be degraded in this case,
but we accept that risk for now. The redundancy will be restored
during SNS repair later.

Relates to https://github.com/Seagate/cortx-motr/issues/1598.

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by andriytk** [comment]
Test result on 1-node setup with 4+2+0 EC pool of 10 disks:

```
$ m0cp -l inet:tcp:192.168.180.165@22501 -H inet:tcp:192.168.180.165@22001 -p 0x7000000000000001:0x53 -P 0x7200000000000001:0x2b -s 16k -c 1 -L 1 /dev/zero -o 0x12345678:0x678900210
motr[454547]:  3a00  ERROR  [io_req_fop.c:363:io_bottom_half]  ioo=0x1328880 from=inet:tcp:192.168.180.165@21003 rc=-5 ti_rc=-5 @<4300000112345678:678900210>
motr[454547]:  3af0   WARN  [io_req.c:1748:ioreq_dgmode_write]  user data written with degraded redundancy: off=0 len=16384 failed_devs=1
$ hctl status -d
Bytecount:
    critical : 0
    damaged : 0
    degraded : 0
    healthy : 27306
Data pool:
    # fid name
    0x6f00000000000001:0x33 'the pool'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x53 'default': 'the pool' None None
Services:
    rocky83  (RC)
    [started]  hax                 0x7200000000000001:0x6          inet:tcp:192.168.180.165@22001
    [started]  confd               0x7200000000000001:0x9          inet:tcp:192.168.180.165@21002
    [started]  ioservice           0x7200000000000001:0xc          inet:tcp:192.168.180.165@21003
    [offline]  m0_client_other     0x7200000000000001:0x2b         inet:tcp:192.168.180.165@22501
    [unknown]  m0_client_other     0x7200000000000001:0x2e         inet:tcp:192.168.180.165@22502
Devices:
    rocky83
    [online]  /dev/null
    [online]  /dev/mapper/bad_disk
    [online]  /dev/loop2
    [online]  /dev/loop3
    [online]  /dev/loop4
    [online]  /dev/loop5
    [online]  /dev/loop6
    [online]  /dev/loop7
    [online]  /dev/loop8
    [online]  /dev/loop9
    [online]  /dev/loop0
$
```

The `bad_disk` was configured like this:

```
$ sudo dmsetup create --concise 'bad_disk,,,,0 18874368 error'
```

**Comment by madhavemuri** [comment]
@andriytk : Any checking being done, if there are more than K failures while writing the data for the parity group ?

**Comment by andriytk** [comment]
@madhavemuri, yes. See the test results in my previous comment (at the bottom).

**Review by madhavemuri — APPROVED** [review]

**Review by nikitadanilov — CHANGES_REQUESTED** [review]

File: motr/io_req.c
Comment (andriytk) [reply motr/io_req.c]: Changed.

Comment (mehjoshi) [reply motr/io_req.c]: I am marking this conversation as resolved because I see the corresponding change request has been addressed in a subsequent commit.

**Comment by mehjoshi** [comment]
@yatin-mahajan , @AbhishekSahaSeagate Can you please review?

**Comment by andriytk** [comment]
@mehjoshi, don't land it please without explicit @nikitadanilov's approval.

**Comment by mehjoshi** [comment]
> @mehjoshi, don't land it please without explicit @nikitadanilov's approval.

Thank you, @andriytk. @chandradharraval kindly note.

**Review by nikitadanilov — APPROVED** [review]

File: motr/io_req.c
Comment (huanghua78) [inline motr/io_req.c:1454]: I think this can be removed.

File: motr/io_nw_xfer.c
Comment (huanghua78) [inline motr/io_nw_xfer.c:1727]: Why is this removed?

So how is the rpc item error populated to "xfer error" and then to "ti" error?

File: motr/io_req.c
Comment (huanghua78) [inline motr/io_req.c:1746]: I am not sure about the condition of "play->pl_attr.pa_S == 0".

If the "play->pl_attr.pa_S" is not zero but the spare units are not used (because SNS repair is has not been triggered), it is still not necessary to continue to do the dg write.

File: motr/io_nw_xfer.c
Comment (andriytk) [reply motr/io_nw_xfer.c:1727]: It's set at `io_bottom_half()`, here: https://github.com/Seagate/cortx-motr/blob/3b70300f63ebd55fb477126b71b01f43326aaa97/motr/io_req_fop.c#L344-L354

File: motr/io_req.c
Comment (andriytk) [reply motr/io_req.c:1454]: We discussed it with @nikitadanilov and agreed that he will look into this jira ticket...

File: motr/io_req.c
Comment (andriytk) [reply motr/io_req.c:1746]: Exactly. This is why we don't skip dg write in this case. Note: this code enables skipping dg write (when we return success to the user, although some units were not written).

Comment (andriytk) [reply motr/io_req.c:1746]: @huanghua78, dg write means writing to the spare units, as far as I understand.

Comment (huanghua78) [reply motr/io_req.c:1746]: I don't agree with you.

A "normal write" means, writing to data units/parity units (if in repaired mode, it may write to spare units because failed units are replaced with spare units). If we have 8+2 layout, then 10 units must *ALL* be written.
A "dg write" means, writing to data units/parity units (again, spare units may involved if in repaired state) but NOT ALL units are written. if we have 8+2 layout, we may write 8 or 9 units.

Comment (andriytk) [reply motr/io_req.c:1746]: Correct. This is exactly what we try to do with this patch if there are no spare units available (which will be the case in our LR/LC configurations). But if there are spare units - old code will try to write into the spare units in dg write mode, and we don't change this logic in this patch.

**Review by huanghua78 — APPROVED** [review]

**Comment by andriytk** [comment]
Started https://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/1195/

**Comment by andriytk** [comment]
Re-started Jenkins test at https://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/1197/.

**Comment by andriytk** [comment]
Started https://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/1200/.

**Comment by andriytk** [comment]
Started https://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/1201/

**Pull request merged**
